### PR TITLE
[Merged by Bors] - Spu smartstream errors

### PR DESCRIPTION
**Pull request by nicholastmosher**
Second sub-piece of https://github.com/infinyon/fluvio/pull/1226

Review/merge after https://github.com/infinyon/fluvio/pull/1326

### CODE REVIEW

**Comment by nicholastmosher** [comment]
@sehz I've updated this next smartstream-error PR, let me know if you have any feedback before merging

File: src/spu/src/services/public/stream_fetch.rs [hunk 1]
```diff
@@ -130,131 +165,133 @@ impl StreamFetchHandler {
...
                             continue;
                         },
                     };
 
-                    if leader_offset_update <= last_consumer_offset {
-                        debug!(ignored_update_offset = leader_offset_update);
-                        last_read_offset = leader_offset_update;
+                    // If the leader offset update is behind the last offset read by
+                    // the consumer, then we still have nothing new to send the consumer.
+                    // Just save the leader offset and continue waiting for events.
```
Comment (sehz) [inline src/spu/src/services/public/stream_fetch.rs:268]: nice comment!

**Review by sehz — CHANGES_REQUESTED** [review]
Overall looks good.  2 comments here

File: src/spu/src/services/public/stream_fetch.rs [hunk 2]
```diff
@@ -876,15 +1192,20 @@ mod test {
...
             topic: topic.to_owned(),
             partition: 0,
             fetch_offset: 0,
             isolation: Isolation::ReadUncommitted,
             max_bytes: 250,
-            wasm_module,
+            wasm_module: Vec::new(),
+            wasm_payload: Some(wasm_payload),
             ..Default::default()
         };
```
[the remark this answers is not in the document]
Comment (nicholastmosher) [reply src/spu/src/services/public/stream_fetch.rs:1211]: Note that this test you commented on is test_filter_stream_max, which is focused on checking max bytes. The other tests do check that if a response is sent, it contains both the correct records and any error that occurred.

File: src/smartstream/derive/src/generator/filter.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
+use quote::quote;
```
Comment (sehz) [inline src/smartstream/derive/src/generator/filter.rs:1]: this is not show stopper but we should reduce duplication

**Review by sehz — APPROVED** [review]
Great job.  This was huge undertaking

**Pull request closed without merging**